### ISSUE TITLE
type checker seems to stop working for certain types after a type where `ts-expect-error` was used

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#45284]
# Bug Report
i spent a good few hours trying to isolate the issue here and i just can't figure it out, so i'll link to a repo where it occurs, but unfortunately it's in a project full of extremely convoluted utility types.

https://github.com/DetachHead/ts-helpers/commit/ea7ff84c29f9c2c925873460d0571c04182f5938#diff-437f36c380b7c6145370d3eb09f4571f486542a4cc0aca11d23f4c855b15519dR334

to summarize, any types that use the `Join` type seem to be ignored if they're used after the `ReplaceValuesWithMap` type is defined.

### steps
1. checkout https://github.com/DetachHead/ts-helpers/tree/nightmare
2. run `tsc`,  the invalid assignment on line 334 is not reported.
3. uncomment line 323
4. run `tsc` again. the invalid assignment is now reported for both assignments

### 🕗 Version & Regression Information
4.4.0-dev.20210802

### ⏯ Playground Link
can't isolate the issue :(
### 💻 Code

```ts
// uncomment this, then the error on the variable assignment below magically comes back
// const foo: Join<['asdf']> = '' // error

/**
 * replaces all instances in `Value` of the first string with the second string with each tuple in `Map`
 * @example
 * type Foo = ReplaceValuesWithMap<'foobarbaz', {foo: 'bar', baz: 'qux'}> // "barbarqux"
 */
export type ReplaceValuesWithMap<Format extends string, Map extends ReplaceValuesMap> =
    // @ts-expect-error stack depth error but it's fine
    Join<_ReplaceValuesWithMap<_TokenizeString<Format, Map>, Map>>

const bar: Join<['asdf']> = '' // no error
```

### 🙁 Actual behavior

no error for invalid assignment on line 334, but th

### 🙂 Expected behavior
same error as on line 323
```
src/utilityTypes/String.ts:323:7 - error TS2322: Type '""' is not assignable to type '"asdf"'.

323 const foo: Join<['asdf']> = '' // error
          ~~~
```

**Comment by RyanCavanaugh** [comment]
```ts
// @ts-expect-error stack depth error but it's fine
```
This is one of those errors that is really not safe to ignore. If we had a notion of errors that are not ignorable, this would be one of them.
